src/s_inter.c

```diff
@@ -74,11 +74,59 @@ that didn't really belong anywhere. */
 #include "pthread.h"
 #endif
 
+#ifdef THREADED_IO
+#include "ringbuffer.h"
+
+#if __STDC_VERSION__ >= 201112L // use stdatomic if C11 is available
+  #include <stdatomic.h>
+  #define SYNC_FETCH(ptr) atomic_fetch_or((_Atomic int *)ptr, 0)
+  #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+          atomic_compare_exchange_strong((_Atomic int *)ptr, &oldval, newval)
+  #define SYNC_STORE(ptr, newval) atomic_store((_Atomic int *)ptr, newval)
+#else // use platform specfics
+  #ifdef __APPLE__ // apple atomics
+    #include <libkern/OSAtomic.h>
+    #define SYNC_FETCH(ptr) OSAtomicOr32Barrier(0, (volatile uint32_t *)ptr)
+    #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+            OSAtomicCompareAndSwap32Barrier(oldval, newval, ptr)
+  #elif defined(_WIN32) || defined(_WIN64) // win api atomics
+    #include <windows.h>
+    #define SYNC_FETCH(ptr) InterlockedOr(ptr, 0)
+    #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+            InterlockedCompareExchange(ptr, oldval, newval)
+  #else // gcc atomics
+    #define SYNC_FETCH(ptr) __sync_fetch_and_or(ptr, 0)
+    #define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
+            __sync_val_compare_and_swap(ptr, oldval, newval)
+  #endif
+  #define SYNC_STORE(ptr, newval) SYNC_COMPARE_AND_SWAP((int*)ptr, (*(int*)ptr), newval)
+#endif
+
+typedef struct _fdsend {
+    int fds_fd;
+    t_fdsendrmfn fds_fn;
+    void *fds_ptr;
+} t_fdsend;
+#endif // THREADED_IO
+
+typedef enum {
+    kFdpManagerAnyThread,
+    kFdpManagerAudioThread,
+#ifdef THREADED_IO
+    kFdpManagerIoThread,
+#endif // THREADED_IO
+} t_fdp_manager;
+
 typedef struct _fdpoll
 {
     int fdp_fd;
     t_fdpollfn fdp_fn;
     void *fdp_ptr;
+    t_fdp_manager fdp_manager;
```

Comment:
Better make this an int, so you don't need the nasty casts in `#define SYNC_STORE(ptr, newval) SYNC_COMPARE_AND_SWAP((int*)ptr, (*(int*)ptr), newval)` :-)

Reply:
and lose the semantic :(

Reply:
> and lose the semantic :(

What to you mean?

Reply:
Ah, you probably mean that the type information is lost. You could just add a comment. The type cast is actually dangerous: `t_fdp_manager` is guaranteed to be 4 bytes and might be padded. When you cast it to an `int *` and dereference it, you would read random data. (In practice this doesn't happen if you allocate with `getbytes`, because it zero initializes the memory.) We already had such an issue: https://github.com/pure-data/pure-data/pull/768

And what if 5 years later some added another struct member of type `char` right after `fdp_manager`?

Reply:
Is 23166e2c an over-engineered solution? I worked on the assumption that C11 atomic will be able to deal with `int` or `enum` just fine.

Reply:
> Is bc0b9fe an over-engineered solution?

I think so ;-)

Personally I would rather add a typedef `t_atomic_int` and define it per platform:

C11: `#define t_atomic_int int`
Apple: `#define t_atomic_int uint32_t`
Win32: `#define t_atomic_int LONG`
GCC: `#define t_atomic_int int`

This has the following advantages:
* it always has the correct size
* people reading code immediately see that it is used for synchronization: `t_atomic_int fdp_manager`.

(`t_atomic_int` could also be used in ringbuffer, btw.)


> I worked on the assumption that C11 atomic will be able to deal with int or enum just fine.

Yes, C11 atomics are type generic, but let's have a look at the definition of `SYNC_COMPARE_AND_SWAP`:

```#define SYNC_COMPARE_AND_SWAP(ptr, oldval, newval) \
          atomic_compare_exchange_strong((_Atomic int *)ptr, &oldval, newval)
  #define SYNC_STORE(ptr, newval) atomic_store((_Atomic int *)ptr, newval)
```

Do you notice the `(_Atomic int *)` cast?

Reply:
f9ad646f63a135760ea7f1691347ece593393b54

Actually it had to be `int32_t` for Apple.

Reply:
> Actually it had to be int32_t for Apple.

Yeah, that's better.

Sidenote: Turns out `OSAtomicOr32Barrier` expects `uint32_t` and `OSAtomicCompareAndSwap32Barrier` expects `int32_t`. We could define `t_atomic_int` as either `int32_t` or `uint32_t`. One of the two methods will always need a pointer cast (which is safe as they have the same size and all relevant modern CPUs implement signed numbers as two's complement :-)

Reply:
Thanks for your patience!

Reply:
>  One of the two methods will always need a pointer cast (which is safe as they have the same size and all relevant modern CPUs implement signed numbers as two's complement :-)

Yes seen that, I left the cast in place for the `uint32_t` :) I am confident if this get merge you'll come around and make all of these atomics more efficient and better written :)